### DISCUSSION TITLE
Non-commercial / hobbyist Licence

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#383] [General]
We're currently discussing whether to add an extra tier for twinBASIC licensing that would sit between the FREE edition (32-bit) and the PROFESSIONAL edition.

The purpose of this extra tier is to accommodate the needs of hobbyists and those that just want to support us on the **_twinBASIC evolution_**, with reduced costs, a slightly reduced feature set and non-commercial licence terms.

This new tier would be the same as the FREE edition, but allow for 64-bit compilation.  To differentiate between the PROFESSIONAL edition, we would NOT include the the optimizing compiler, and the main difference would be that the licence terms would disallow any commercial use of binaries generated with the compiler.

The intended pricing for the new tier would be GBP £15/month, reduced to £7.50/month in the discounted pre-order offer.

Would there be much interest in this?

**Comment by dmrvb** [comment]
I'm not sure about this.
£15 a month is a lot for a hobby.
I suspect most hobby/non commercial users will be using TB as a VB6 rather than VBA  (I could well be wrong, what do you think is the likely VB6/VBA split?) and the free version will do most of what is needed. Is 64 bit compilation actually that important for VB6? Most Apps run fine as 32bit though this might change in the future. 
Im probably not typical but for me the optimised compiler is more attractive than 64 bit.
Something else I would pay for is some better controls, but I am not sure if you should be doing controls or hoping that the community will provide them?
Its probably too late but an extra charge for some of the new "non VB" additions like the longlong might be an option.

**Comment by WaynePhillipsEA** [comment]
Thanks for your thoughts on this @dmrvb.   The reason I started this thread was because we've been asked several times over the past month for a cheaper option with 64-bit support, largely from developers that don't do professional / commercial development work.

From my personal point of view, I absolutely agree with you.  We created the free 32-bit tier precisely for those developers that class themselves as hobbyists / non-professionals.

Since we've been asked several times, I figured I'd move the discussion onto here to see what others think.  This extra tier certainly wont be added unless there's enough call for it, so we'll see how it goes.

**Comment by Kr00l** [comment]
As @dmrvb pointed out, LongLong is a great motivation even on 32 bit tier only. So restricting the LongLong from the free version and make available earliest on the hobby license is probably an idea.

**Comment by Hojo-Norem** [comment]
I've been keeping an eye on twinBASIC for a while now and in all honesty, as somebody who codes for a hobby in fits and starts, my heart sank when I discovered that twinBASIC was going for a subscription model.

Now, I don't think I'd have trouble convincing myself to drop a chunk of change on a feature complete 'drop-in replacement for VB6 pro' twinBASIC Pro v1.0 because no matter how long it takes, I _will_ get my moneys worth out of it.

**Reply by WaynePhillipsEA** [reply]
Hi, and thanks for joining the discussion!

Firstly, I have to admit to be being totally out of touch when it comes to hobbyist programming.  The main reason for us choosing the subscription model is to try to ensure twinBASIC has a long and vibrant future, but trying to look at it from a hobbyist point of view, where (I guess) you might not use the tool every month... I can see it might be difficult to justify a regular expense.

My first question to you (and anyone else who also classes themselves as a hobbyist) is...  is there a price point where the subscription model would not be off-putting?

**Reply by Hojo-Norem** [reply]
TBO, to my own fault I can't come up with a price point.  I don't want to suggest something that sounds insulting.

Reading the other suggestions, if tB gains traction then the nag-screen idea may be ineffectual as I can imagine that it would probably be simple for somebody to figure out some form of auto patcher to remove them.

I had some of my own ideas, like locking compiled DLLs and EXEs to the machine that compiled them and locking the compiler down to a single account which in turn is locked down to a single machine...  A bit long winded, I know.  Probably not really viable and just as easy to circumvent as the nag-screens unless you encrypt the binaries.

**Comment by mansellan** [comment]
Just throwing an idea in there, what about having the compiler embed a pop up a banner into the binaries when using a non-commercial / hobbyist license (except on Win32, which is already free)? That should prevent any license abuse by commercial usages. Not sure how that could work for DLLs though.

Would that help keep the non-commercial license cost down?

**Reply by WaynePhillipsEA** [reply]
Thanks @mansellan, that could be a better option than restricting the feature set too much.

**Comment by dmrvb** [comment]
I also don't like the subscription model but I think the really big thing here is that we can possibly have some influence on how TB develops and that's worth paying a bit of money for.
Is the end of August deadline going to extend or is it time to pay???

**Reply by WaynePhillipsEA** [reply]
Thanks @dmrvb.  With regard to the August deadline, we are extending the offer for a further three months, but reducing the discount by 5% each month.  e.g. September orders will get 45% lifetime discount, October 40%, November 35%.

Once the Professional edition is released, the offer will be renamed 'early-bird discount' instead of 'pre-order'.

**Comment by YogiYang** [comment]
Hello @WaynePhillipsEA ,

Personally I feel this is (hobbyist version) just not required. Mind you it will surely get misused. You are already giving the standard edition for free.

I don't get it as to why would a person need support for 64bit when working on a hobby project.

I feel a hobbyist should build the app in the Standard edition and test thoroughly. Once s/he is satisfied purchase the Professional edition and compile the code for getting final compiled exe in 64bit.

**Comment by FullValueRider** [comment]
If you want twinBasic to be successful outside of those who have professional application or other commitments to VB/x then you need to ensure that the barriers to entry are lower than the available alternatives.  This essentially means that to compete with C#/VB Net or Type/Javascipt etc, any form of a fee is likely to be a killer.  

Consequently I would suggest that you allow your biggest and best twinBasic environment to be used for free and find an alternative method for restricting the use of the compiler output.  Once such method could be a constraint that without a paid for license, twinBasic output would be restricted to run only on the combination of internal compiler Id/computer/office installation that generated the code.

If you see twinBasic being a key part to enhancing desktop Office applications then for many potential users the availability of both 32 and 64 bit versions would be critical, as many of your potential hobbyist/starter users won't have control over which version of Office they have installed.

**Reply by WaynePhillipsEA** [reply]
Thanks @FullValueRider.  You raise very good points about 32/64 bit support... the lack of availability of the 64-bit version would certainly hinder adoption with many VBA developers that might want to try tB generated DLLs inside their projects.   This needs some thought, but a machine-based restricted compilation might indeed be a good solution.

**Reply by dmrvb** [reply]
I have used a few third party controls, and much much prefer to have full functionality available with nag screens in the IDE and compiled EXE. This allows me to learn how to use the software and fully evaluate it before spending money. Any other arrangement is a significant deterrent.

Nag screen should be no more than an irritation to hobby users.

There are a couple of potential big issues:.........
All users will generate some support effort so the free users are loss makers, and unhappy unsupported users posting bad reviews need to be avoided, so having more users paying just a small amount _Might_ be a good way to go_
Piracy is another problem and making full functionality freely available might be an additional risk, though the baddies will find a way regardless. TB will be a powerful compiler of unmanaged code so will generate some interest from the writers of malware, and other software low-life.  I believe that PurBasic had this problem.

**Reply by mwolfe02** [reply]
@dmrvb Some very good points that I hadn't considered such as unsupported users posting bad reviews and TB drawing the interest of malware writers.  Not much you can do about the latter, but the former is something that deserves some thought and attention.

**Reply by WaynePhillipsEA** [reply]
In my experience I've found that the more effort you put into trying to prevent pirating simply results in the pirates being more interested and keen to crack your software... they seem to feed off the challenge.  So whilst I will obviously put some form of protection in place (such as a nag screen or machine-lock), I won't be investing vast amounts of time in it, as my time is better spent elsewhere.  If someone decides to use a cracked version of tB from some unknown source with the high risks of malware being included with it, then so be it.

The support issue for the free users may well become an issue... I guess time will tell on that one.  Hopefully a community will evolve that helps each other out to some extent.

**Comment by Greedquest** [comment]
**tl;dr** We should totally have a few more things in the free license because VBA and VB6 people have different requirements and IMO the licensing currently appeals mostly to VB6 needs and isn't sufficient for VBA hobbyists or new starters. 


What @FullValueRider suggests in https://github.com/WaynePhillipsEA/twinbasic/discussions/383#discussioncomment-1282166 is similar to my line of thinking, I'll just add a few extra points:

Firstly, I think if twinBasic wishes to migrate users it must not charge for those features which already exist freely in alternatives. For example, VB6 does not offer 64-bit compilation and so that can be considered a valuable unique selling point (USP) of tB. However, VBA users are used to 64-bit support, as you say, and so this feature of tB offers no exciting draw and is not something you can market on. Conversely VBA has no way to create a standalone .exe while VB6 does, so twinBasic .exe support is an "exciter" to VBA consumers but a "satisfier" to VB6 consumers. Factors like execution speed, compiled dll size, even the editor experience are what I'd call "performance" features - i.e. the better their performance (relative to alternatives) the more valuable they are to both VBA and VB6 users. You can split all of tB's features into these 3 categories of "exciter/order-winner", "satisfier/required" and "performance/the more the better", and you'll find that there are subtle differences depending on whether you look at it from the perspective of a VBA, VB6 or in fact a user from another background entirely. 

Specifically talking about 64-bit tB support - Office (and by extension VBA) is now 64-bit by default. Arguably VBA is also a growing/sustainable language whereas VB6 is shrinking (this is going on the data for language usage by the SO developer survey, which has seen a reduction in VB6, and it's no longer listed in the top 40 most used languages, VBA is number 22 in 2021). Perhaps also new users of a language means newer hardware means 64-bit OS and Office. Finally, a marginally smaller fraction of VBA users are professional than VB6 users (again SO developer survey, 2017) so they may be less likely to shell out for a subscription to support their hobby.* All this is to say I think VBA users for whom 64-bit is expected rather than a USP/order-winner are also likely to represent a larger proportion of future potential twinBasic users and are also less likely to be able to afford a professional license in general/ see the value in the current feature set. They might donate to the project though... 

FWIW this is how I would classify the feature set (I think I've got most of them) according to appeal to different user types (but I'm not a user of VB6 so I may be off the mark on some of them):

| Feature                                                                             | Satisfier     | Performance    | Exciter          |
|-------------------------------------------------------------------------------------|---------------|----------------|------------------|
| 64-bit                                                                              | VBA           |                | VB6              |
| Compilation to EXE                                                                  | VB6           |                | VBA              |
| Cross-Platform                                                                      | VBA (partial) |                | VB6              |
| Office Integration                                                                  | VBA           | VB6            | New-Starters     |
| Multithreading language support                                                     |               |                | VB*              |
| Inheritance                                                                         |               |                | VB*              |
| Package manager                                                                     |               |   VB*             |               |
| VSCode as editor                                                                    |               |        VB*        |               |
| VB* Syntax                                                                          | VB*           |                | New-Starters     |
| Optimised Compiling                                                                 |               | VB*            |                  |
| Analysis Extension API                                                              |               | VB*            |                  |
| [Zero runtime requirements](https://nolongerset.com/twinbasic-runtime-requirements/) |               | VB6 (tiny dll) | VBA/New-Starters |


Note I've included "New-Starters" in the table - that's because another angle on all of this is new coder adoption as opposed to existing migration. A fairly standard model, especially in software where the market can be extremely crowded, since barrier to entry is pretty low (not to downplay the complexity of what tB is doing, but there are a lot of programming languages out there), is to offer a free and practically feature-complete version to students and new adopters to get them hooked. For example, think GitHub offering free private 3-contributor repositories so students or small teams can work on projects without sharing their code online or unlimited members in public repos for open-source projects. By appealing to users early on when they're students or hobbyists they ensure that when some fraction of those users progress into industry they are already familiar with the GitHub workflow and so request it in their new jobs, which is when they are charged for GitHub Pro. GitHub needs to offer so much of its product free so that it can set itself apart from BitBucket or any other git host, because the market is crowded and if it offered a limited free product when better free products exist, even if its paid version is far superior to alternatives, people will already be used to working with the competitor and many will just stick to that when they go professional. Programming languages like twinBasic in particular are skills that take practice, so if you fail to capture users early it's very hard to get them back as they will be entrenched in the alternative products.

A final point to note; I think you mentioned in some post that you might be looking to crowdsource controls for your GUI or crowdsource some standard library functions. It would be particularly nice if some of the compatibility features like multi-platform support or multi-bitness support could be included free for contributors to the tB ecosystem, so they can test shared code. Creating things for others to use it what drives me most and part of that involves testing my code on multiple architectures so I know it can work for the maximum number of people.

---

*<sub> A lot of caveats in that SO developer survey but my intuition, maybe I'm wrong, does say that VB6 users are more likely to be coming from industry, VBA users contain more of a mix - and certainly I think that is the future trend - how many new users are likely to be adopting VB6 as a hobby rather than a newer language? </sub>

**Reply by mansellan** [reply]
So this seems to indicate that it would be sensible to allow 64-bit use, exclusively for VBA use, under the free license. Obviously that's a decision only Wayne can make, but I'm wondering if it's technically possible? Just disallowing 64-bit .exe's may not be robust enough - a DLL is only an entry point and a PE header away from an .exe after all. Perhaps it would need to restrict to only running in known VBA hosts?

**Comment by jimekus** [comment]
To produce a groundswell of support, enough to attract the paying users, I suggest dual licensing for free and unfettered personal use - separated from paid commercial users. Fully restrictive licensing surrounding closed source, that lets the market whittle away the passion projects - leaving only commercial code, is worse than you can imagine. See Niakwa Basic-2C licensing death.

**Comment by WaynePhillipsEA** [comment]
We've decided (for now at least), that we will allow 64-bit compilation for non-licence holders, though EXE and DLL compilations will be subject to an embedded twinBASIC splash screen, displayed for 5 seconds at startup.  This feature is currently active in v0.13.1, now that it has been released.

**Reply by FullValueRider** [reply]
Hurrah

**Reply by YogiYang** [reply]
That is really great news. 👍

**Comment by mansellan** [comment]
Have you considered offering free licenses to (carefully vetted) OSS projects? It could help drive adoption - you could perhaps embed a subtle watermark into the executable to watch for abuse, and require such projects to attribute "compiled with twinBASIC" on their project sites.

Worked example - [Rubberduck](https://rubberduckvba.com/) is [considering implementing](https://github.com/rubberduck-vba/Rubberduck/issues/5176) an LSP front-end/back-end split. twinBASIC would be a fantastic fit for the front-end - we've had issues from day 1 caused by using managed code to interact with the VBE COM model over COM-interop. But at the same time, it would be unthinkable to require all contributors to purchase a twinBASIC license to run the build, and unworkable to suggest that contributors only use 32-bit VBA hosts to develop and test locally. We're volunteers...

But if contributors could build and test against 32 and 64 bit hosts (putting up with the splash screen on 64-bit during local dev), safe in the knowledge that final GitHub green builds would be built by a sanctioned copy of twinBASIC that lacked the splash, well.... that could work!

**Reply by Greedquest** [reply]
I don't know how the full license works exactly, but couldn't that workflow be achieved with a single paid license just for the GitHub green CI/CD builds? Or does the RD release build happen on a contributor's machine under their licence and they just upload the release?

Just playing devil's advocate here as I think a free license for OSS, particularly small projects where even a single license is prohibitively expensive, could be justified as a reasonable tradeoff of immediate sales against nurturing and growing the userbase.

**Reply by YogiYang** [reply]
> Have you considered offering free licenses to (carefully vetted) OSS projects? It could help drive adoption - you could perhaps embed a subtle watermark into the executable to watch for abuse, and require such projects to attribute "compiled with twinBASIC" on their project sites.
> 
> Worked example - [Rubberduck](https://rubberduckvba.com/) is [considering implementing](https://github.com/rubberduck-vba/Rubberduck/issues/5176) an LSP front-end/back-end split. twinBASIC would be a fantastic fit for the front-end - we've had issues from day 1 caused by using managed code to interact with the VBE COM model over COM-interop. But at the same time, it would be unthinkable to require all contributors to purchase a twinBASIC license to run the build, and unworkable to suggest that contributors only use 32-bit VBA hosts to develop and test locally. We're volunteers...
> 
> But if contributors could build and test against 32 and 64 bit hosts (putting up with the splash screen on 64-bit during local dev), safe in the knowledge that final GitHub green builds would be built by a sanctioned copy of twinBASIC that lacked the splash, well.... that could work!

Instead of providing OSS project license to all OSS project developers it would be better if @WaynePhillipsEA personally decides as to whom to provide free license as his discretion is a better idea.

Of course this will also increase his work but at lease he will have complete controls as to who will get free license and who will not!

**Reply by WaynePhillipsEA** [reply]
I'd certainly be happy to donate a licence to the RD project.  Unfortunately I don't know much about AppVeyor, but assuming it can be configured to launch a custom executable, then I see no reason why it can't be achieved.  

Although at the moment tB doesn't offer command-line compilation (opened #508), plus I'm guessing you'll want to wait until you can create an ActiveX control for making tool windows.

**Reply by mansellan** [reply]
> Instead of providing OSS project license to all OSS project developers it would be better if @WaynePhillipsEA personally decides as to whom to provide free license as his discretion is a better idea.
> 
> Of course this will also increase his work but at lease he will have complete controls as to who will get free license and who will not!

Yes sorry that's what I meant by the "carefully vetted" bit - sorry I wasn't clear enough. The license would only be at Wayne's discretion, and just be a single instance for CI use only, not for individual developers.

@WaynePhillipsEA I don't have much to do with the AV side either, and at the moment this is just an idea on my part. Probably best to wait until this is confirmed as the direction wanted for RD, and (if so) we are ready for CI - the existing licenses are fine for us to undertake a proof of concept. Not to mention that we'd need to wait for the UI side too. I was just sounding out whether this would be a path you'd want to take.

**Comment by jimekus** [comment]
In my estimation, getting ready to relaunch OSS VB6 projects, based on
highly mature working code, most likely requires integrating vbRichClient
as a work-in-process, and can be first done using only the VB6 IDE.

I developed a mongrel tableau studio using an outdated set of AI primitives
with an artistic flair It eventually became My DJ Reader that I use for
several hours a day.

The artistic part uses DirectX as well as GDI+ and intrinsic VB6 graphics.
It is the latter that could benefit from decoupling code. I could easily do
such a port, unless the following idea takes hold.

I would therefore like to evaluate a single TwinBasic educational/school
licence with the aim of vetting such integration projects, and as
motivation to teach apprentices. Done well, adding stuff like Cairo could
be farmed out to students without any experience.

**Comment by vbBaka** [comment]
I think the free edition should be:
- 32 & 64bit
- basic components (to allow backward compatible)
- limited activex support
- optimized (to show its fully power)
- multithreading 
- splash screen to tell "this is done in TwinBasic" of course I would like the community to help creating a new screen.
- u get only major-updates (maybe once every 6 months or once a year)

the pro edition:
- additional components (and future adds)
- extended activex support
- additional tools, to do all kinds of stuff (analysis)
- office integration and other more "serious business" stuff
- built-in helper (like msdn library) so u have all the help u need
- inheritance 
- package manager
- crossplatform
- u get all updates so always up to date.

the reason u should give more to the free edition, is to show its potential.
if someone new is looking around for a programming language and see that this is only 32bit, it could scare him away.
the same with optimization (not sure what this means exactly) and multithreading. this to show that person its a powerful programming language 2022 and not 2005.

the pro-edition should always feel like "this is to help u develop faster, by giving u assistance, tools, frequent updates, crossplatform, and all kinds of stuff that are useful for someone working in a office environment"
I believe in a "patron" model. u "donate" as much u can. if 1$ or 100$. of course the one donating 1$ will not get everything.
like in gaming community, the 100$ will get the release first, he could get a "u can decide the next monster we gonna add" etc.
the 50$ maybe gets the beta release. and could be part of different poll, to decide stuff about the game etc.
while the 1$ will need to wait for last. 
even so, there could be 1000-1$ and that would generate 1000$ so its not bad to have a lot of 1 dollars, that dont have that much money but want to contribute a bit.
